clippy_lints/src/needless_pass_by_value.rs

```diff
@@ -0,0 +1,303 @@
+
+            // Determines whether `ty` implements `Borrow<U>` (U != ty) specifically.
+            // This is needed due to the `Borrow<T> for T` blanket impl.
+            let implements_borrow_trait = preds.iter()
+                .filter_map(|pred| if let ty::Predicate::Trait(ref poly_trait_ref) = *pred {
+                    Some(poly_trait_ref.skip_binder())
+                } else {
+                    None
+                })
+                .filter(|tpred| tpred.def_id() == borrow_trait && &tpred.self_ty() == ty)
+                .any(|tpred| &tpred.input_types().nth(1).expect("Borrow trait must have an parameter") != ty);
+
+            if_let_chain! {[
+                !is_self(arg),
+                !ty.is_mutable_pointer(),
+                !is_copy(cx, ty, node_id),
+                !implements_trait(cx, ty, fn_trait, &[], Some(node_id)),
+                !implements_trait(cx, ty, asref_trait, &[], Some(node_id)),
+                !implements_borrow_trait,
+
+                let PatKind::Binding(mode, defid, ..) = arg.pat.node,
+                !moved_vars.contains(&defid),
+            ], {
+                // Note: `toplevel_ref_arg` warns if `BindByRef`
+                let m = match mode {
+                    BindingMode::BindByRef(m) | BindingMode::BindByValue(m) => m,
+                };
+                if m == Mutability::MutMutable {
+                    continue;
+                }
+
+                // Suggestion logic
+                let sugg = |db: &mut DiagnosticBuilder| {
+                    if_let_chain! {[
+                        match_type(cx, ty, &paths::VEC),
+                        let TyPath(QPath::Resolved(_, ref path)) = input.node,
+                        let Some(elem_ty) = path.segments.iter()
+                            .find(|seg| &*seg.name.as_str() == "Vec")
+                            .map(|ps| ps.parameters.types()[0]),
+                    ], {
+                        let slice_ty = format!("&[{}]", snippet(cx, elem_ty.span, "_"));
+                        db.span_suggestion(input.span,
+                                        &format!("consider changing the type to `{}`", slice_ty),
+                                        slice_ty);
+                        return;
+                    }}
+
+                    if match_type(cx, ty, &paths::STRING) {
+                        db.span_suggestion(input.span,
+                                           "consider changing the type to `&str`",
+                                           "&str".to_string());
+                    } else {
+                        db.span_suggestion(input.span,
+                                           "consider taking a reference instead",
+                                           format!("&{}", snippet(cx, input.span, "_")));
+                    }
+
+                    // Suggests adding `*` to dereference the added reference.
+                    if let Some(spans) = spans_need_deref.get(&defid) {
```

Comment:
this isn't happening in the `Vec` case anymore.

It would probably be best if it (both the derefs and the main suggestion) were a single suggestion with multiple spans and replacements. Use [`util::multispan_sugg`](https://github.com/Manishearth/rust-clippy/blob/9a7597d5af4d882ddc92bf568a1c7ba582bb4b55/clippy_lints/src/utils/mod.rs#L570) to create such a suggestion.